python-bindings/chiapos.cpp

```diff
@@ -77,23 +77,39 @@ PYBIND11_MODULE(chiapos, m)
 
     py::class_<DiskProver>(m, "DiskProver")
         .def(py::init<const std::string &>())
+        .def(py::pickle(
+            [](const DiskProver &dp) { // __getstate__
+                return py::make_tuple(dp.GetFilename(),
+                                      dp.GetSize(),
+                                      dp.GetMemo(),
+                                      dp.GetId(),
+                                      dp.GetTableBeginPointers(),
+                                      dp.GetC2());
+                },
+                [](const py::tuple& t) { // __setstate__
+                if (t.size() != 6)
+                    throw std::runtime_error("Invalid state!");
+
+                auto filename = t[0].cast<std::string>();
+                auto k = t[1].cast<uint8_t>();
+                auto memo = t[2].cast<std::vector<uint8_t>>();
+                auto id = t[3].cast<std::vector<uint8_t>>();
+                auto table_begin_pointers = t[4].cast<std::vector<uint64_t>>();
```

Comment:
I have a feeling that these are offsets into the plot file, and that we should check that they are indeed all smaller than the size of the plot file. We've had many problems where we crash on corrupt or truncated plot files in the past.

Reply:
I was thinking about this now for a while and i was even about to just us the normal filename constructor like in [here](https://github.com/Chia-Network/chiapos/compare/18136f09ec1c55d5be3f418b1d244ecc0d7d3f37..438ff5e98747a5399f81be80639a9f91d9bf2878) but now after i just pushed the change it came to my mind that this will cause trouble if we depend on the file while de-serializing if the file does not longer exist.

Reply:
What *should* happen if the file doesn't exist? Today we don't populate the plot list with the file to begin with. Throwing an exception, I suppose, would stop loading of all other plots...

Reply:
Well, it shouldn't fail to de-serialize the whole cached data if only one plot file is not longer available but this will happen if we throw in the constructor because the file doesn't exist. The idea was to just serialize the whole list of plots and load it on startup like in https://github.com/xdustinface/chia-blockchain/commit/2ca417ab548e9a34fec43b27c1ed3ca0009259d0. The `PlotManager` would then take care of removed plots in is refresh task. With your proposed change we would need to store a cache file for each plot.

Reply:
I'm only proposing this case is handled well (and maybe even tested). But also the case where the pickled file is corrupt or the plot is corrupt.

Reply:
I'm don't get what you want/mean here with the corrupt pickle/plot file as this only adds support for pickle. Last push added basic validation and a comment to point out that the new constructor assumes the parameter match the underlying file. Any thoughts/objections?

Reply:
is it possible to violate the class invariants via the constructor you use when un-pickling?

Reply:
Obviously, like the comment says you can provide parameter that doesn't match the underlying file but so can you load a corrupt plot without knowing with the other constructor, right? I could of course add some `void EnsureValidated()` method which checks a `fIsValidated` flag before each read to verify the file pointers are in bounds if that flag is false but at the and this doesn't improve the constructor and we use `SafeRead` everywhere which *should* throw if something doesn't match and all what may happen if the pickled file for some weird reason i can't think about gets altered in a way that a pointer exceeds the bounds nothing bad can happen besides a failing lookup or do i miss something here?

Reply:
> Obviously, like the comment says you can provide parameter that doesn't match the underlying file but so can you load a corrupt plot without knowing with the other constructor, right?

I don't know. I'm not very familiar with this class. It sounds like it doesn't have very many invariants then.

> all what may happen if the pickled file for some weird reason i can't think about gets altered in a way that a pointer exceeds the bounds nothing bad can happen besides a failing lookup or do i miss something here?

It's hard to know without a test.

Reply:
The last two commits and this https://github.com/Chia-Network/chia-blockchain/pull/8009/files#diff-1c950ad1eb2a79b51cb4f61cf39b8fe8d09d28ad963ce7dd24107053ff316d14R113-R126 should take care of this. Any objections or what else do you want to see tested related to this?